File: swift/Tests/RezelTests/CasesTests.swift
import Testing
import Foundation
@testable import Rezel

private let caseDir = URL(fileURLWithPath: #filePath)
    .deletingLastPathComponent()
    .deletingLastPathComponent()
    .deletingLastPathComponent()
    .deletingLastPathComponent()
    .appendingPathComponent("web")
    .appendingPathComponent("test")
    .appendingPathComponent("cases")

private func externalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
    if name == "ext1" {
        return ExternalTokenizer({ input, _ in
            if input.next == Int(Character("{").asciiValue!) {
                input.advance()
                input.acceptToken(terms["braceOpen"]!)
            } else if input.next == Int(Character("}").asciiValue!) {
                input.advance()
                input.acceptToken(terms["braceClose"]!)
            } else if input.next == Int(Character(".").asciiValue!) {
                input.advance()
                input.acceptToken(terms["Dot"]!)
            }
        })
    }
    fatalError("Undefined external tokenizer \(name)")
}

private func externalSpecializer(name: String, terms: [String: Int]) -> (String, Stack) -> Int {
    if name == "spec1" {
        return { value, _ in
            if value == "one" { return terms["one"]! }
            if value == "two" { return terms["two"]! }
            return -1
        }
    }
    fatalError("Undefined external specialize \(name)")
}

private func externalPropFn(_ name: String) -> NodePropBase {
    return NodeProp<String>(deserialize: { x in x })
}

private struct CaseFile {
    let name: String
    let grammar: String
    let testContent: String
    let expectedError: String?

    init?(fileName: String, content: String) {
        let name = (fileName as NSString).deletingPathExtension
        self.name = name

        if let range = content.range(of: "\n# ") {
            grammar = String(content[content.startIndex..<range.lowerBound])
            testContent = String(content[range.lowerBound...])
        } else {
            grammar = content
            testContent = ""
        }

        if let errRange = grammar.range(of: "//! ", options: .backwards) {
            expectedError = String(grammar[errRange.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        } else {
            expectedError = nil
        }
    }
}

@Suite("Cases", .serialized)
struct CasesTests {
    private static let skipSet: Set<String> = []

    @Test func cases() throws {
        let fm = FileManager.default
        let files = try fm.contentsOfDirectory(at: caseDir, includingPropertiesForKeys: nil)
        var caseFiles: [CaseFile] = []
        for fileURL in files {
            guard fileURL.pathExtension == "txt" else { continue }
            let content = try String(contentsOf: fileURL, encoding: .utf8)

            guard let cf = CaseFile(fileName: fileURL.lastPathComponent, content: content) else { continue }
            caseFiles.append(cf)
        }

        caseFiles.sort { $0.name < $1.name }

        for cf in caseFiles {
            let noCases = cf.testContent.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            if noCases { continue }
            if cf.expectedError != nil { continue }
            if Self.skipSet.contains(cf.name) { continue }

            let fileURL = caseDir.appendingPathComponent(cf.name + ".txt")
            let parser = try buildParser(cf.grammar, options: BuildOptions(
                fileName: fileURL.path,
                externalTokenizer: externalTokenizer,
                externalSpecializer: externalSpecializer,
                externalProp: externalPropFn
            ))
            let tests = try fileTests(cf.testContent, cf.name + ".txt")
            for t in tests {
                do {
                    try t.run(parser)
                } catch {
                    Issue.record("\(cf.name)/\(t.name): \(error)")
                }
            }
        }
    }
}

File: dotnet/Rezel.Tests/CasesTests.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using Rezel.Common;
using Rezel.Generator;
using Rezel.Lr;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using File = System.IO.File;

namespace Rezel.Tests;

[TestClass]
public class CasesTests
{
    static readonly string CaseDir = Path.Combine(
        Path.GetDirectoryName(typeof(CasesTests).Assembly.Location)!,
        "..", "..", "..", "..", "..", "web", "test", "cases");

    static ExternalTokenizer? ExternalTokenizerFn(string name, Dictionary<string, int> terms)
    {
        if (name == "ext1")
            return new ExternalTokenizer((input, stack) =>
            {
                var next = input.Next;
                if (next == '{')
                {
                    input.Advance();
                    input.AcceptToken(terms["braceOpen"]);
                }
                else if (next == '}')
                {
                    input.Advance();
                    input.AcceptToken(terms["braceClose"]);
                }
                else if (next == '.')
                {
                    input.Advance();
                    input.AcceptToken(terms["Dot"]);
                }
            });
        return null;
    }

    static Func<string, Stack, int>? ExternalSpecializerFn(string name, Dictionary<string, int> terms)
    {
        if (name == "spec1")
            return (value, stack) =>
            {
                if (value == "one") return terms["one"];
                if (value == "two") return terms["two"];
                return -1;
            };
        return null;
    }

    static NodePropBase? ExternalPropFn(string name)
    {
        if (name == "tag")
            return new NodeProp<string>(x => x);
        return null;
    }

    static string ExtractGrammar(string content, out string rest)
    {
        var idx = content.IndexOf("\n# ", StringComparison.Ordinal);
        if (idx >= 0)
        {
            rest = content[(idx + 1)..];
            return content[..(idx + 1)];
        }
        rest = "";
        return content;
    }

    public static IEnumerable<object[]> GetCaseFiles()
    {
        if (!Directory.Exists(CaseDir)) yield break;
        foreach (var file in Directory.GetFiles(CaseDir, "*.txt"))
        {
            yield return new object[] { file };
        }
    }

    [TestMethod]
    [DynamicData(nameof(GetCaseFiles))]
    public void RunGrammarTestCase(string filePath)
    {
        var fileName = Path.GetFileName(filePath);
        var content = File.ReadAllText(filePath);
        var grammar = ExtractGrammar(content, out var rest);

        var expectedErrMatch = Regex.Match(grammar, @"//! (.*)");
        var hasCases = Regex.IsMatch(rest, @"\S");

        if (expectedErrMatch.Success)
        {
            var expectedErr = expectedErrMatch.Groups[1].Value.Trim().ToLower();
            Exception? caught = null;
            try
            {
                BuildExt.BuildParser(grammar, new BuildOptions
                {
                    FileName = filePath,
                    ExternalTokenizerFn = ExternalTokenizerFn,
                    ExternalSpecializer = ExternalSpecializerFn,
                    ExternalProp = ExternalPropFn,
                    Warn = e => throw new InvalidOperationException(e),
                });
            }
            catch (Exception e)
            {
                caught = e;
            }
            Assert.IsNotNull(caught, $"Expected exception containing '{expectedErr}'");
            StringAssert.Contains(caught.Message.ToLower().Replace("\\u0022", "\\\""), expectedErr);
            if (!hasCases) return;
        }

        Assert.IsTrue(hasCases, $"Test with neither expected errors nor input cases ({fileName})");

        var tests = TestUtils.FileTests(rest, fileName);
        var parser = BuildExt.BuildParser(grammar, new BuildOptions
        {
            FileName = filePath,
            ExternalTokenizerFn = ExternalTokenizerFn,
            ExternalSpecializer = ExternalSpecializerFn,
            ExternalProp = ExternalPropFn,
            Warn = e => throw new InvalidOperationException(e),
        });

        foreach (var test in tests)
        {
            test.Run(parser);
        }
    }
}
